File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/2096-내려가기/solution.py
import sys
si = sys.stdin.readline

n = int(si())
a= [0] * 4
dy_min = [[10000000, 10000000, 10000000, 10000000] for _ in range(2)]
dy_max = [[0, 0, 0, 0] for _ in range(2)]

# 초기값 채우기
a = [0] + list(map(int, si().split()))
for j in range(1, 4):
    dy_min[1][j] = a[j]
    dy_max[1][j] = a[j]

# 점화식을 토대로 dy 배열 채우기
for i in range(2, n + 1):
    a = [0] + list(map(int, si().split()))
    for cur in range(1, 4):
        dy_min[i % 2][cur] = 10000000
        dy_max[i % 2][cur] = 0
        for prev in range(1, 4):
            if abs(cur - prev) > 1:
                continue
            dy_min[i % 2][cur] = min(dy_min[i % 2][cur], dy_min[(i + 1) % 2][prev] + a[cur])
            dy_max[i % 2][cur] = max(dy_max[i % 2][cur], dy_max[(i + 1) % 2][prev] + a[cur])

print(max(dy_max[n % 2][1:]), end=' ')
print(min(dy_min[n % 2][1:]))

File: 강의 자료/02-알고리즘/15~17-동적 프로그래밍/문제별 코드/2096-내려가기/solution.java
import java.io.*;
import java.util.*;
import java.lang.Math;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N;
    static int[][] dy_min, dy_max;
    static int[] A;
    
    static void input(){
        N = scan.nextInt();
        A = new int[4];
        dy_min = new int[2][4];
        dy_max = new int[2][4];
    }

    static void pro() {
        // 초기값 구하기
        for (int j = 1; j <= 3; j++) {
            dy_min[1][j] = scan.nextInt();
            dy_max[1][j] = dy_min[1][j];
        }

        // 점화식을 토대로 Dy 배열 채우기
        for (int i = 2; i <= N; i++) {
            for (int j = 1; j <= 3; j++) A[j] = scan.nextInt();
            for (int cur = 1; cur <= 3; cur++) {
                dy_max[i % 2][cur] = 0;
                dy_min[i % 2][cur] = 10000000;
                for (int prev = 1; prev <= 3; prev++) {
                    if (Math.abs(cur - prev) > 1) continue;
                    dy_max[i % 2][cur] = Math.max(dy_max[i % 2][cur], dy_max[(i + 1) % 2][prev] + A[cur]);
                    dy_min[i % 2][cur] = Math.min(dy_min[i % 2][cur], dy_min[(i + 1) % 2][prev] + A[cur]);
                }
            }
        }
        System.out.print(Math.max(Math.max(dy_max[N % 2][1], dy_max[N % 2][2]), dy_max[N % 2][3]) + " ");
	    System.out.print(Math.min(Math.min(dy_min[N % 2][1], dy_min[N % 2][2]), dy_min[N % 2][3]) + " ");
    }

    public static void main(String[] args) {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}